From d661513f0f11190ffc1397847b65c9b9c922d337 Mon Sep 17 00:00:00 2001
From: Masaharu Hayakawa <masaharu.hayakawa.ry@renesas.com>
Date: Wed, 11 Oct 2017 13:22:07 +0900
Subject: [PATCH 209/909] mmc: core: Add timeout errors to retuning condition

commit f6fef5ca01cd423bbf02c7b4f407be2815c5e310 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

When conditions such as temperature change, timeout may
occur due to communication failure with SD/MMC.

In case of timeout error, the re-tuning flag is not true.
Therefore, the command is retried without re-tuning.
If re-tuning is executed, the success rate of retries
can be improved.

By adding timeout to retuning conditions, re-tuning is
executed when retrying.

Signed-off-by: Masaharu Hayakawa <masaharu.hayakawa.ry@renesas.com>
[saito: rework commit message.]
Signed-off-by: Takeshi Saito <takeshi.saito.xv@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mmc/core/core.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/core/core.c b/drivers/mmc/core/core.c
index 1061ec010f92..ba1847d47540 100644
--- a/drivers/mmc/core/core.c
+++ b/drivers/mmc/core/core.c
@@ -143,11 +143,13 @@ void mmc_request_done(struct mmc_host *host, struct mmc_request *mrq)
 	struct mmc_command *cmd = mrq->cmd;
 	int err = cmd->error;
 
-	/* Flag re-tuning needed on CRC errors */
+	/* Flag re-tuning needed on CMD, CRC, END errors and timeoout */
 	if ((cmd->opcode != MMC_SEND_TUNING_BLOCK &&
 	    cmd->opcode != MMC_SEND_TUNING_BLOCK_HS200) &&
-	    (err == -EILSEQ || (mrq->sbc && mrq->sbc->error == -EILSEQ) ||
+	    (err == -EILSEQ || err == -ETIMEDOUT ||
+	    (mrq->sbc && mrq->sbc->error == -EILSEQ) ||
 	    (mrq->data && mrq->data->error == -EILSEQ) ||
+	    (mrq->data && mrq->data->error == -ETIMEDOUT) ||
 	    (mrq->stop && mrq->stop->error == -EILSEQ)))
 		mmc_retune_needed(host);
 
-- 
2.17.1

